### PR TITLE
Refactor argument parser using cmdopt

### PR DESCRIPTION
**Pull request by mohanson**
Modified part of the source code of cmdopt, the diff results are as follows

```
2,5d1
<  * Copied from https://bellard.org/nncp/nncp-2023-10-21.tar.gz
<  */
< 
< /*
7c3
<  *
---
>  * 
29d24
< #include "my_stdlib.h"
31d25
< #include "my_stdio.h"
34,35c28,30
< #include "my_string.h"
< #include "my_assert.h"
---
> #include <assert.h>
> #include "cutils.h"
> 
57c52
< 
---
>     
71c66
< 
---
>     
110,111d104
< /* See the reason: https://reviews.llvm.org/D103009 */
< #pragma clang optimize off
121d113
< #pragma clang optimize on
140,141c132,133
<     arg_pos = 0;
<     for(optind = 0; optind < argc; ) {
---
>     arg_pos = 1;
>     for(optind = 1; optind < argc; ) {
217c209
< #endif
---
> #endif    
225c217
<     size_t len;
---
>     size_t len, i;
227c219
< 
---
>     
230d221
<         char l[256] = {0};
240c231
<                 strcat(l, " ");
---
>                 putchar(' ');
243c234
<             strcat(l, "-");
---
>             putchar('-');
245c236
<                 strcat(l, "-");
---
>                 putchar('-');
248c239,240
<             memcpy(&l[col+1], p, len);
---
>             for(i = 0; i < len; i++)
>                 putchar(p[i]);
250c242
< 
---
>             
255c247
< 
---
>         
257,264c249,252
<             if (d->arg_desc) {
<                 strcat(l, " ");
<                 strcat(l, d->arg_desc);
<                 col += 1 + strlen(d->arg_desc);
<             } else {
<                 strcat(l, " arg");
<                 col += 4;
<             }
---
>             if (d->arg_desc)
>                 col += printf(" %s", d->arg_desc);
>             else
>                 col += printf(" arg");
272c260
<             strcat(l, " ");
---
>             putchar(' ');
275,276c263
<         strcat(l, d->desc);
<         printf("%s", l);
---
>         printf("%s\n", d->desc);
302c289
< 
---
>     
332c319
< 
---
>     
348c335
< 
---
> 
```

### CODE REVIEW

File: quickjs/qjs.c [hunk 1]
```diff
@@ -273,6 +224,39 @@ static int run_from_cell_data(JSContext *ctx, bool enable_fs) {
...
+    int err = 0;
+    size_t buf_size = 0;
+    size_t index = 0;
+
+    err = load_cell_code_info_explicit(&buf_size, &index, code_hash, hash_type);
+    if (err) {
+        return err;
+    }
+
+    char buf[buf_size + 1];
```
Comment (XuJiandong) [inline quickjs/qjs.c:247]: Use malloc to avoid stack overflow issue.

File: quickjs/qjs.c [hunk 2]
```diff
@@ -285,18 +269,36 @@ static JSContext *JS_NewCustomContext(JSRuntime *rt) {
     return ctx;
 }
 
+static const CMDOptDesc nncp_options[] = {
```
Comment (XuJiandong) [inline quickjs/qjs.c:272]: rename "nncp_options" to "js_vm_options"

File: quickjs/qjs.c [hunk 3]
```diff
@@ -311,50 +313,65 @@ int main(int argc, const char **argv) {
...
-            JS_SetModuleLoaderFunc(rt, NULL, js_module_dummy_loader, NULL);
-            err = compile_from_file(ctx);
-            break;
-        default:
-            printf("unknow type: %d", type);
-            return -1;
+    if (cmdopt_has(co, "c")) {
+        JS_SetModuleLoaderFunc(rt, NULL, js_module_dummy_loader, NULL);
+        err = compile_from_file(ctx);
+        return err;
```
Comment (XuJiandong) [inline quickjs/qjs.c:323]: Use `CHECK`, don't return immediately.

File: quickjs/qjs.c [hunk 4]
```diff
@@ -311,50 +313,65 @@ int main(int argc, const char **argv) {
...
+            js_std_loop(ctx);
+        }
+        return err;
+    }
+
+    err = run_from_cell_data(ctx, false);
+    if (err == 0) {
+        js_std_loop(ctx);
+    }
+    return err;
```
Comment (XuJiandong) [inline quickjs/qjs.c:373]: No returning here.

File: quickjs/qjs.c [hunk 5]
```diff
@@ -311,50 +313,65 @@ int main(int argc, const char **argv) {
...
+        JS_SetModuleLoaderFunc(rt, NULL, js_module_dummy_loader, NULL);
+        err = compile_from_file(ctx);
+        return err;
     }
-    CHECK(err);
+    if (cmdopt_get(co, "e")) {
+        const char *data = cmdopt_get(co, "e");
+        err = eval_buf(ctx, data, strlen(data), "<cmdline>", 0);
+        if (err == 0) {
+            js_std_loop(ctx);
```
Comment (XuJiandong) [inline quickjs/qjs.c:329]: Can we move this call js_std_loop to later?

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: quickjs/qjs.c (hunk 5 above)
Comment (mohanson) [reply quickjs/qjs.c:343]: js_std_loop is not needed when compiling bytecode, I prefer to put it in the branch where it is needed.

Comment (XuJiandong) [reply quickjs/qjs.c:343]: I think, js_std_loop can be used while compiling.

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**